lib/elixir/lib/kernel.ex

```diff
@@ -4981,6 +4981,28 @@ defmodule Kernel do
     module = :elixir_aliases.concat([env.module, h])
     alias = String.to_atom("Elixir." <> Atom.to_string(h))
 
+    case :lists.keyfind(alias, 1, env.aliases) do
+      # If the alias is the same as the module being defined, it means we're redefining
+      # the same nested module. In that case, let's not warn specifically about "shadowed"
+      # aliases here.
+      {^alias, ^module} ->
+        :ok
+
+      {^alias, other_parent} ->
+        defmod_name = Enum.join([h | t], ".")
+
+        IO.warn(
+          "the nested call to \"defmodule #{defmod_name}\" will not implicitly alias " <>
+            "#{inspect(alias)} here because there is already an alias #{inspect(other_parent)} " <>
+            "that resolves to the same alias. To fix this, either remove the " <>
+            "#{inspect(other_parent)} alias, or define the nested module outside of the parent",
```

Comment:
I am trying to think about the error message. What do you think about:

    you are defining a module named Child.Comment, inside module Post,
    but the module will be defined as RootChild.Comment instead of Post.Child
    due to the alias Child which points to RootChild. You must address this ambiguity
    either by removing the alias Child or by defining the module outside of the parent

Reply:
I think that's not what happens though, because there is technically no ambiguity, is just that we don't set up an implicit alias. I updated the error message, I like it a lot better now. If you're good with it, I'll merge.